pageserver/src/repository.rs

```diff
@@ -66,60 +65,103 @@ pub trait Repository: Send + Sync {
+    Idle,
 }
 
-/// A state of the timeline synchronization with the remote storage.
-/// Contains `disk_consistent_lsn` of the corresponding remote timeline (latest checkpoint's disk_consistent_lsn).
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
-pub enum TimelineSyncState {
-    /// No further downloads from the remote storage are needed.
-    /// The timeline state is up-to-date or ahead of the remote storage one,
-    /// ready to be used in any pageserver operation.
-    Ready(Lsn),
+pub enum LocalTimelineState {
+    // timeline is loaded into memory (with layer map and all the bits),
+    Loaded(LoadedLocalTimelineState),
+    // timeline is on disk locally and ready to be loaded into memory.
+    Unloaded,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
+pub enum RemoteTimelineState {
     /// Timeline is scheduled for downloading, but its current local state is not up to date with the remote storage.
     /// The timeline is not ready to be used in any pageserver operations, otherwise it might diverge its local state from the remote version,
     /// making it impossible to sync it further.
     AwaitsDownload(Lsn),
     /// Timeline was not in the pageserver's local working directory, but was found on the remote storage, ready to be downloaded.
     /// Cannot be used in any pageserver operations due to complete absence locally.
     CloudOnly(Lsn),
+}
+
+/// A state of the timeline synchronization with the remote storage.
+/// Contains `disk_consistent_lsn` of the corresponding remote timeline (latest checkpoint's disk_consistent_lsn).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
+pub enum TimelineState {
+    /// No further downloads from the remote storage are needed.
+    /// The timeline state is up-to-date or ahead of the remote storage one,
+    /// ready to be used in any pageserver operation.
+    Local(LocalTimelineState),
+
+    /// Timeline data is not available locally, download might be scheduled.
+    Remote(RemoteTimelineState),
+
     /// Timeline was evicted from the pageserver's local working directory due to conflicting remote and local states or too many errors during the synchronization.
     /// Such timelines cannot have their state synchronized further and may not have the data about remote timeline's disk_consistent_lsn, since eviction may happen
     /// due to errors before the remote timeline contents is known.
     Evicted(Option<Lsn>),
 }
 
-impl TimelineSyncState {
-    pub fn remote_disk_consistent_lsn(&self) -> Option<Lsn> {
-        Some(match self {
-            TimelineSyncState::Evicted(None) => return None,
-            TimelineSyncState::Ready(lsn) => lsn,
-            TimelineSyncState::AwaitsDownload(lsn) => lsn,
-            TimelineSyncState::CloudOnly(lsn) => lsn,
-            TimelineSyncState::Evicted(Some(lsn)) => lsn,
-        })
-        .copied()
+impl From<RepositoryTimeline> for TimelineState {
```

Comment:
NIT: feels a bit odd to _move_ an entire timeline to get its state, as if it could be a method on the `RepositoryTimeline` instead.

Reply:
What do you think about implementing From for a reference? 

Something like that:
```rust
struct Foo {
    a: u32
}

struct Bar {
    b: u32
}

impl<'a> From<&'a Foo> for Bar {
    fn from(_: &Foo) -> Self { todo!() }
}
```

Maybe usual method is more readable

Reply:
No objections against lifetimes per se, so why not.
A small concern is a readability of that `From` usage, but we can always replace it with a plain function in case of any issues.